### ISSUE TITLE
Add Sleep Activity

### ISSUE THREAD

**Issue by cpfeiffer** [Freeyourgadget/Gadgetbridge#45]
The user should be able to get a visualization of his sleep phases. He should be able to see the time when he fell asleep, the sleep phases and when he woke up.

Ideally there should be a visualization of sleep history, but maybe that should be handled by a service as proposed in #49

**Comment by ashimokawa** [comment]
To visualize sleep data, morpheuz seems to send sleep data to the authors homepage which is also open source. And we could just integrate this in Gadgetbridge to visalize everything offline inside a webview.

https://github.com/JamesFowler42/morpheuz20/tree/master/website/morpheuz

To get sleep data, my plan is to interface with morpheuz on the pebble watch directly without using its javascript code that is supposed to run inside the original Pebble App.

The miband data could be converted in the same format before sending it to our webview rendered diagrams.

**Comment by cpfeiffer** [comment]
Anyone who wants to figure out how to get sleep data from the Mi Band or the Pebble?

**Comment by cpfeiffer** [comment]
See http://forum.xda-developers.com/general/accessories/xiaomi-mi-band-data-extraction-t3019156 for some research on the original Mi Fit app's data (sqlite based).

If we could create the same db from the data available via the BLE API, we could use the same extraction logic.

**Comment by ashimokawa** [comment]
Sleep data for the Pebble is Pebble App specific.
For Morpheuz it should be easy since it is open source.
I should be able to do it.

**Comment by ashimokawa** [comment]
I will hack an initial prototype which displays morpheuz data.

**Comment by danielegobbetti** [comment]
I've got my first 11 bytes of activity data from the miband! I think sleep and steps data are all part of these, I'll keep you posted.

**Comment by cpfeiffer** [comment]
Whoohoo! :-) Looking very much forward to that!

**Comment by ashimokawa** [comment]
@danielegobbetti 
Great news.  :+1: 
If you have more details, let's think about how to unify these data with what I get from Morpheuz in a common format for storing it in a SQLite database.

**Comment by danielegobbetti** [comment]
What i know sp far:
The mi band sends data with a resolution of one minute.
The recived data has to be acked and only then the miband will send the data of the following minute. 

I still have not decoded the data unfortunately, so that's what i know for now.

**Comment by cpfeiffer** [comment]
Cool, can't wait to have a look at that :-)

**Comment by danielegobbetti** [comment]
Hello again, I pushed a branch https://github.com/danielegobbetti/Gadgetbridge/tree/get_stats because I need help.

when reading from the control point MiBandService.UUID_CHARACTERISTIC_ACTIVITY_DATA I only get 11 bytes of data. I alread made sense of those, but it's only a timestamp + a series of counters of how much further data have to be read. The problem is that I cannot read those further data :(

I think we may need to listen for notifications on that UUID, and I tried to add a NotifyAction in the btle package, but my method is not getting called. Can you please take a look at my code and tell me if you see something obviously wrong/missing?

On the plus side, I know how to acknowledge the data! On the minus side I emptied the memory of my band and I'll have to do some activity! :)

**Comment by cpfeiffer** [comment]
Cool, I'll have a look. Enjoy some fitness action in the meantime :-)

**Comment by danielegobbetti** [comment]
Thanks a lot!

Don't worry about my code on the btle stack, if can implement working notify code I'll take care of the conflicts / merge on my branch!

Thanks!

**Comment by cpfeiffer** [comment]
OK, for a start, I can reproduce the behavior. The code looks good, it's just that the MI Band doesn't notify us of new data.

I have shifted some code to send acks also in different occasions, but that didn't help yet.

Still trying...

**Comment by danielegobbetti** [comment]
If you use the "read" method instead of notify do you get 11 bytes as well?

**Comment by cpfeiffer** [comment]
Yes.

BTW, I think we need to enable notifications not just for the entire characteristic, but for one or more specific values (specified by UUID). Don't know which, yet.

**Comment by danielegobbetti** [comment]
I may be making some progress... Apparentrly it's not enough to do setCharateristicNotification

**Comment by cpfeiffer** [comment]
Yes. You have to get some properties of descriptor for uuid 2902... and enable notifications with that as well. Or at least it looks like that. The only difference I have is that I cannot read any activity data at all anymore ;-)

**Comment by cpfeiffer** [comment]
Also it seems that there are a few well-known characteristics for which notifications are enabled once on startup. Maybe we need more notifications and react to changes on them.

**Comment by danielegobbetti** [comment]
I see what you did :)
FYI: If you acknowledge with a status != 0 the band will not "consume" the data it has on record.

I guess we are looking at the same code ;-)

**Comment by cpfeiffer** [comment]
:-D

I'm trying to run the notify-methods on initialize() now.

**Comment by danielegobbetti** [comment]
Oh... I'm still trying to solve how to write to a descriptor using the methods in GadgetBridge

**Comment by cpfeiffer** [comment]
I'm wondering what 2902 and 2901 are...

**Comment by cpfeiffer** [comment]
This is my current WIP:

        boolean result = false;
        if (gatt.setCharacteristicNotification(getCharacteristic(), true)) {
            BluetoothGattDescriptor sleepDescriptor = getCharacteristic().getDescriptor(MiBandService.UUID_CHARACTERISTIC_SLEEP_DATA);
            if (sleepDescriptor != null) {
                int properties = getCharacteristic().getProperties();
                if ((properties & 0x10) > 0) {
                    LOG.info("properties & 0x10 > 0");
                    sleepDescriptor.setValue(BluetoothGattDescriptor.ENABLE_NOTIFICATION_VALUE);
                    result = gatt.writeDescriptor(sleepDescriptor);
                } else if ((properties & 0x20) > 0){
                    LOG.info("properties & 0x20 > 0");
                    sleepDescriptor.setValue(BluetoothGattDescriptor.ENABLE_INDICATION_VALUE);
                    result = gatt.writeDescriptor(sleepDescriptor);
                }
            } else {
                LOG.warn("sleep descriptor null");
            }
        } else {
            LOG.error("Unable to enable notification for " + getCharacteristic().getUuid());
        }
        return result;

**Comment by cpfeiffer** [comment]
sleepDescriptor is 2902 (for lack of a better name). It's probably something completely different.

**Comment by danielegobbetti** [comment]
2902 seems standard: https://developer.bluetooth.org/gatt/descriptors/Pages/DescriptorViewer.aspx?u=org.bluetooth.descriptor.gatt.client_characteristic_configuration.xml

**Comment by danielegobbetti** [comment]
Thanks for the code!

**Comment by cpfeiffer** [comment]
(This is the NotifyAction.run() method)

**Comment by ashimokawa** [comment]
Maybe we should rename this issue to something Mi Band specific? ;)

**Comment by cpfeiffer** [comment]
Maybe we should find a chatroom instead of a ticket system ;)

**Comment by ashimokawa** [comment]
gitter.im?

**Comment by danielegobbetti** [comment]
Sorry for the noise @ashimokawa ! Do you have a jabber account @cpfeiffer ?

**Comment by cpfeiffer** [comment]
Interesting thing about 2902. BTLE is weird. So this 2902 is a standard characteristic inside a specific UUID namespace (0000*2902*-0000-1000-8000-00805f9b34fb)

**Comment by cpfeiffer** [comment]
@danielegobbetti Yes. I just create a chatroom gadgetbridge@conference.jabber.ccc.de

**Comment by danielegobbetti** [comment]
I'm in gitter.im

**Comment by danielegobbetti** [comment]
Ok I'll join the chatroom! (disregard the gitter message)

**Comment by cpfeiffer** [comment]
FWIW; I have now created a gadgetbridge room on gitter.im as well

**Comment by cpfeiffer** [comment]
The current state works pretty good for me. During initialization, I enable notifications, which works realiably now. Reading of the activities also works reliably.

Can you test the current state of the branch? Next stop: a separate activity for that.

**Comment by danielegobbetti** [comment]
I'm getting data as well! +1

**Comment by 1337sup3rh4x0r** [comment]
Seems you guys had a productive night! Looking forward to the next release!

**Comment by danielegobbetti** [comment]
Hi, i added some logic to deal (correctly) with the data received through notification. If I do a pull request I create a duplicate of #55 because that commit is in the history of my branch, hence I am writing here to please use this commit https://github.com/danielegobbetti/Gadgetbridge/commit/118a014af2eb42f050fe765b2d4bba9026bf5813 :)

**Comment by cpfeiffer** [comment]
Done. I cherry-picked that commit (and the next one), thanks!

**Comment by cpfeiffer** [comment]
@1337sup3rh4x0r  Sorry for the spamming yesterday, we somehow managed to use this ticket as a chatroo ;-)

**Comment by 1337sup3rh4x0r** [comment]
No need to apologize, I am the leacher here :)

**Comment by danielegobbetti** [comment]
@ashimokawa, speaking of the data format, are there requirements (a data model) that Morpheuz is imposing?
I am asking because from the MiBand we are getting three bytes per minute, and these raw data are always present, no matter what the activity is (sleeping, running, etc.). Every other format we need should be calculated over this raw data, hence I believe we may use the same model of Morpheuz and aggregate the MiBand data in the same way.

I think I found the part of the code that interprets the raw data, but it does so according to the need of the original apps.

**Comment by ashimokawa** [comment]
Morpheuz seems to work in two modes.
1) When Bluetooth is not connected it stores one value for each 10 minutes. (0-5000) which indicates "movement". It records for 9 hours of sleep which is 54 values in total stored as a short (16 bit) value internally. These values can be transferred afterwards one by one, each one which has to be acknowledged. 
2) When connected to the phone it transmits one value of activity data per minute and transfers it to the phone immediately . I do not know yet if these values are accumulated each minute for a set of 10 samples or if I would have to accumulate these myself.

In both cases eventually 54 samples of sleep data is sent to a website which  draws nice diagrams.
It would not be hard to see how we have to interpret the data since the website and watchapp code is open source.

But it is not an easy decision if I should
1) support morpheuz because it is highly regarded in the pebble community
2) write my own app which extends the scope of morpheuz by also collecting non-sleep activity data for 24 hours or so, improve the way data is sent the phone in terms of speed/efficiency, and adapt to the format you are using on the Mi Band.

**Comment by cpfeiffer** [comment]
@danielegobbetti do you want to add your work to @ashimokawa's db support and sleep activity or shall I do that? I would simply merge your work into master and keep on working there.

**Comment by danielegobbetti** [comment]
I'll do it later or early tomorrow. I did some refactoring today that I haven't pushed yet!

**Comment by cpfeiffer** [comment]
Great, looking forward to that :-)

**Comment by danielegobbetti** [comment]
I pushed to my fork and created a PR for you to review.

I also added a comment (github comment, not code comment) wrt getting the data into the database. This works but can be done only after merging with master to get the great DB functionality that @ashimokawa added. I can push it after this is done.

**Comment by cpfeiffer** [comment]
Darn. Missed a hashmark in my commit 8aef92026c73593d515616770b3e43389515c5c7
so it did not automatically show up here.

**Comment by chrisvella** [comment]
Attempting to view the sleep monitor crashes the entire app for me. I am using the latest master branch. Here is my log file:

```
 21:30:20.600 [main] DEBUG n.f.g.BluetoothCommunicationService - Service startcommand: nodomain.freeyourgadget.gadgetbride.bluetoothcommunicationservice.action.connect
21:30:20.601 [main] INFO  n.f.g.b.BtLEQueue - Disconnecting BtLEQueue from GATT device
21:30:20.605 [GadgetBridge GATT Dispatcher] DEBUG n.f.g.b.BtLEQueue - Thread interrupted
21:30:20.605 [GadgetBridge GATT Dispatcher] INFO  n.f.g.b.BtLEQueue - Queue Dispatch Thread terminated.
21:30:20.609 [main] INFO  n.f.g.b.BtLEQueue - Attempting to connect to MI
21:30:20.618 [main] INFO  n.f.g.GB - Setting broadcast receivers to: true
21:30:20.634 [main] INFO  n.f.g.GB - Setting broadcast receivers to: true
21:30:20.653 [GadgetBridge GATT Dispatcher] DEBUG n.f.g.b.BtLEQueue - Queue Dispatch Thread started.
21:30:23.051 [Binder_1] DEBUG n.f.g.b.BtLEQueue - connection state change, newState: 2 (success)
21:30:23.078 [Binder_1] INFO  n.f.g.b.BtLEQueue - Connected to GATT server.
21:30:23.104 [Binder_1] INFO  n.f.g.b.BtLEQueue - Attempting to start service discovery:true
21:30:23.114 [main] INFO  n.f.g.ControlCenter - device connected, requesting more info
21:30:23.115 [Binder_1] INFO  n.f.g.m.MiBandSupport - Attempting to pair MI device...
21:30:23.120 [Binder_1] DEBUG n.f.g.m.MiBandSupport - Writing User Info!
21:30:23.121 [Binder_1] DEBUG n.f.g.m.MiBandSupport - Requesting Battery Info!
21:30:23.121 [Binder_1] DEBUG n.f.g.b.BtLEQueue - about to add: Transaction task: Initializing device with 11 actions
21:30:23.122 [main] INFO  n.f.g.GB - Setting broadcast receivers to: true
21:30:23.136 [main] INFO  n.f.g.GB - Setting broadcast receivers to: true
21:30:23.151 [main] DEBUG n.f.g.BluetoothCommunicationService - Service startcommand: nodomain.freeyourgadget.gadgetbride.bluetoothcommunicationservice.action.request_versioninfo
21:30:23.152 [main] INFO  n.f.g.m.MiBandSupport - Attempting to pair MI device...
21:30:23.154 [main] DEBUG n.f.g.m.MiBandSupport - Writing User Info!
21:30:23.155 [main] DEBUG n.f.g.m.MiBandSupport - Requesting Battery Info!
21:30:23.156 [main] DEBUG n.f.g.b.BtLEQueue - about to add: Transaction task: Initialize device with 12 actions
21:30:23.156 [main] DEBUG n.f.g.b.BtLEQueue - about to add: Transaction task: Get MI Band device info with 1 actions
21:30:23.430 [Binder_2] DEBUG n.f.g.b.BtLEQueue - characteristic write: 0000ff0f-0000-1000-8000-00805f9b34fb (success)
21:30:23.431 [Binder_2] INFO  n.f.g.m.MiBandSupport - Successfully paired  MI device
21:30:23.528 [Binder_1] DEBUG n.f.g.b.BtLEQueue - characteristic write: 0000ff04-0000-1000-8000-00805f9b34fb (success)
21:30:23.530 [GadgetBridge GATT Dispatcher] DEBUG n.f.g.b.TransactionBuilder - use NOTIFICATION
21:30:23.534 [main] INFO  n.f.g.ControlCenter - device connected, requesting more info
21:30:23.535 [main] INFO  n.f.g.GB - Setting broadcast receivers to: true
21:30:23.556 [main] DEBUG n.f.g.BluetoothCommunicationService - Service startcommand: nodomain.freeyourgadget.gadgetbride.bluetoothcommunicationservice.action.request_versioninfo
21:30:23.559 [main] DEBUG n.f.g.b.BtLEQueue - about to add: Transaction task: Get MI Band device info with 1 actions
21:30:23.626 [Binder_1] DEBUG n.f.g.b.BtLEQueue - descriptor write: 00002902-0000-1000-8000-00805f9b34fb (success)
21:30:23.631 [GadgetBridge GATT Dispatcher] DEBUG n.f.g.b.TransactionBuilder - use NOTIFICATION
21:30:23.722 [Binder_3] DEBUG n.f.g.b.BtLEQueue - descriptor write: 00002902-0000-1000-8000-00805f9b34fb (success)
21:30:23.723 [GadgetBridge GATT Dispatcher] DEBUG n.f.g.b.TransactionBuilder - use NOTIFICATION
21:30:23.820 [Binder_2] DEBUG n.f.g.b.BtLEQueue - descriptor write: 00002902-0000-1000-8000-00805f9b34fb (success)
21:30:23.821 [GadgetBridge GATT Dispatcher] DEBUG n.f.g.b.TransactionBuilder - use NOTIFICATION
21:30:23.919 [Binder_1] DEBUG n.f.g.b.BtLEQueue - descriptor write: 00002902-0000-1000-8000-00805f9b34fb (success)
21:30:23.924 [GadgetBridge GATT Dispatcher] DEBUG n.f.g.b.TransactionBuilder - use NOTIFICATION
21:30:24.017 [Binder_3] DEBUG n.f.g.b.BtLEQueue - descriptor write: 00002902-0000-1000-8000-00805f9b34fb (success)
21:30:24.115 [Binder_2] DEBUG n.f.g.b.BtLEQueue - characteristic write: 0000ff0a-0000-1000-8000-00805f9b34fb (success)
21:30:24.212 [Binder_1] DEBUG n.f.g.b.BtLEQueue - characteristic read: 0000ff0c-0000-1000-8000-00805f9b34fb (success)
21:30:24.216 [GadgetBridge GATT Dispatcher] INFO  n.f.g.b.CheckInitializedAction - Aborting device initialization, because already initialized: nodomain.freeyourgadget.gadgetbridge.GBDevice@9bdf4f36
21:30:24.221 [GadgetBridge GATT Dispatcher] ERROR n.f.g.b.BtLEQueue - Action returned false: CheckInitializedAction on characteristic: (null)
21:30:24.235 [main] INFO  n.f.g.ControlCenter - device connected, requesting more info
21:30:24.240 [main] INFO  n.f.g.GB - Setting broadcast receivers to: true
21:30:24.303 [main] INFO  n.f.g.ControlCenter - device connected, requesting more info
21:30:24.308 [main] INFO  n.f.g.GB - Setting broadcast receivers to: true
21:30:24.310 [Binder_1] DEBUG n.f.g.b.BtLEQueue - characteristic read: 0000ff01-0000-1000-8000-00805f9b34fb (success)
21:30:24.376 [main] INFO  n.f.g.GB - Setting broadcast receivers to: true
21:30:24.407 [Binder_3] DEBUG n.f.g.b.BtLEQueue - characteristic read: 0000ff01-0000-1000-8000-00805f9b34fb (success)
21:30:24.450 [main] INFO  n.f.g.GB - Setting broadcast receivers to: true
21:30:24.485 [main] DEBUG n.f.g.BluetoothCommunicationService - Service startcommand: nodomain.freeyourgadget.gadgetbride.bluetoothcommunicationservice.action.request_versioninfo
21:30:24.486 [main] DEBUG n.f.g.BluetoothCommunicationService - Service startcommand: nodomain.freeyourgadget.gadgetbride.bluetoothcommunicationservice.action.request_versioninfo
21:30:24.491 [main] INFO  n.f.g.GB - Setting broadcast receivers to: true
21:30:24.504 [main] INFO  n.f.g.GB - Setting broadcast receivers to: true
21:30:30.874 [main] DEBUG n.f.g.BluetoothCommunicationService - Service startcommand: nodomain.freeyourgadget.gadgetbride.bluetoothcommunicationservice.action.notification_sms
21:30:30.881 [main] INFO  n.f.g.m.MiBandSupport - Sending notification to MiBand: android.bluetooth.BluetoothGattCharacteristic@c125c80
21:30:30.881 [main] DEBUG n.f.g.b.BtLEQueue - about to add: Transaction task: sms received with 12 actions
21:30:30.937 [Binder_1] DEBUG n.f.g.b.BtLEQueue - characteristic write: 0000ff05-0000-1000-8000-00805f9b34fb (success)
21:30:30.939 [Binder_1] INFO  n.f.g.m.MiBandSupport - handleControlPoint got status:0
21:30:30.940 [Binder_1] INFO  n.f.g.m.MiBandSupport - handleControlPoint GOT DATA:0x       8
21:30:30.940 [Binder_1] INFO  n.f.g.m.MiBandSupport - handleControlPoint GOT DATA:0x       1
21:30:30.940 [Binder_1] DEBUG n.f.g.b.BtLEQueue - characteristic changed: 0000ff03-0000-1000-8000-00805f9b34fb
21:30:31.524 [Binder_2] DEBUG n.f.g.b.BtLEQueue - characteristic write: 0000ff05-0000-1000-8000-00805f9b34fb (success)
21:30:31.527 [Binder_2] INFO  n.f.g.m.MiBandSupport - handleControlPoint got status:0
21:30:31.530 [Binder_2] INFO  n.f.g.m.MiBandSupport - handleControlPoint GOT DATA:0x      13
21:30:32.107 [Binder_3] DEBUG n.f.g.b.BtLEQueue - characteristic write: 0000ff05-0000-1000-8000-00805f9b34fb (success)
21:30:32.108 [Binder_3] INFO  n.f.g.m.MiBandSupport - handleControlPoint got status:0
21:30:32.109 [Binder_3] INFO  n.f.g.m.MiBandSupport - handleControlPoint GOT DATA:0x       8
21:30:32.109 [Binder_3] INFO  n.f.g.m.MiBandSupport - handleControlPoint GOT DATA:0x       1
21:30:32.109 [Binder_3] DEBUG n.f.g.b.BtLEQueue - characteristic changed: 0000ff03-0000-1000-8000-00805f9b34fb
21:30:32.694 [Binder_1] DEBUG n.f.g.b.BtLEQueue - characteristic write: 0000ff05-0000-1000-8000-00805f9b34fb (success)
21:30:32.700 [Binder_1] INFO  n.f.g.m.MiBandSupport - handleControlPoint got status:0
21:30:32.702 [Binder_1] INFO  n.f.g.m.MiBandSupport - handleControlPoint GOT DATA:0x      13
21:30:33.280 [Binder_2] DEBUG n.f.g.b.BtLEQueue - characteristic write: 0000ff05-0000-1000-8000-00805f9b34fb (success)
21:30:33.283 [Binder_2] INFO  n.f.g.m.MiBandSupport - handleControlPoint got status:0
21:30:33.287 [Binder_2] INFO  n.f.g.m.MiBandSupport - handleControlPoint GOT DATA:0x       8
21:30:33.292 [Binder_2] INFO  n.f.g.m.MiBandSupport - handleControlPoint GOT DATA:0x       1
21:30:33.300 [Binder_2] DEBUG n.f.g.b.BtLEQueue - characteristic changed: 0000ff03-0000-1000-8000-00805f9b34fb
21:30:33.864 [Binder_2] DEBUG n.f.g.b.BtLEQueue - characteristic write: 0000ff05-0000-1000-8000-00805f9b34fb (success)
21:30:33.868 [Binder_2] INFO  n.f.g.m.MiBandSupport - handleControlPoint got status:0
21:30:33.871 [Binder_2] INFO  n.f.g.m.MiBandSupport - handleControlPoint GOT DATA:0x      13
21:30:42.245 [main] DEBUG n.f.g.BluetoothCommunicationService - BluetoothCommunicationService is being created
21:30:42.248 [main] INFO  n.f.g.BluetoothCommunicationService - no intent
```

**Comment by ashimokawa** [comment]
This should be some SQL syntax problem in table creation, right @danielegobbetti ?

**Comment by danielegobbetti** [comment]
Yes I've got the same issue with a fresh install and the problem was SQL related on creation of the db. This should happen only on android >5 and I hope to fix it later today.

You should be getting an SQL syntax error in the logcat though, that I can't see in the OP, but I'm on mobile and I might have overlooked it.

--

**Comment by ashimokawa** [comment]
@danielegobbetti 
Happens to me on Kitkat also after deleting all data and forcing recreation of the db

**Comment by chrisvella** [comment]
I have never used sleep tracking before, and the official app is removed. So I assume it would make sense the db being empty.

**Comment by danielegobbetti** [comment]
The db is ours, the official app wouldn't have helped in any case :) 

But I'm happy I'm not the only one who never used the official app! ;)
--

**Comment by danielegobbetti** [comment]
Good to know, hope to fix it soon!
--

**Comment by danielegobbetti** [comment]
It should be fixed on master now.

**Comment by chrisvella** [comment]
I can confirm the fix works @danielegobbetti!
The connect was lost when fetching the activity data for the first time, but I have been unable to replicate it (so I can't give you any logs).
This is really cool. Thank you for developing the fix.
Thank you @ashimokawa for looking into the issue too.

**Comment by danielegobbetti** [comment]
In case it happens again and you're able to capture logs please head to this issue https://github.com/Freeyourgadget/Gadgetbridge/issues/68 which is more specific.
Thanks for confirming the fix works and for the kind words :)

--

**Comment by danielegobbetti** [comment]
@cpfeiffer the new sleep activity is much nicer! 

Since there is still the issue that the number of steps at night is close to 0, the graph at night is mostly empty. I had a proposal to make it better:
- use intensity (instead of steps) for bars height
- plot the number of steps using a linegraph (evt. with a second y axis)

What do you think?

**Comment by cpfeiffer** [comment]
@danielegobbetti AFAICS we do use intensity for all but sleep activity already. Do you know the range of intensity? Maybe the movement_divisor (256) is wrong or we should simply increase the values a bit?

Does the number of steps make sense for sleep data at all? Do you want the steps in the same chart or in a secondary one?

**Comment by chrisvella** [comment]
Just built master with your changes @cpfeiffer. The sleep monitor makes a lot more sense now. Good work everyone! This feature is coming along nicely. 

I have noticed the time at the beginning and end of the sleep period, overlap with overall time values (from ____ to ____). Might be an idea to raise them. 
Also its not immediately obvious that you can manipulate the graph view. A heading under the title might be good.

**Comment by danielegobbetti** [comment]
@cpfeiffer regarding the deep/light sleep issue I believe we are still far from the solution. Looking at different nights there are some where the current setting makes sense, and other nights where the opposite is true. I think we need to collect data from as many bands as possible to analyze them and finding what the true interpretation is.
The same is true for intensity, 256 was set by me as a rule of thumb.

I assumed the number of steps was used for everything in the graph because of the boolean that was set, I did not look at the code :)

In any case the last additions to the graph are really nice!

**Comment by 1337sup3rh4x0r** [comment]
can you tell me how to provide my tracking data to you guys to help figuring this out? I have the current version from FDroid installed

**Comment by danielegobbetti** [comment]
You could do an adb backup and extract the database from it, but you need a computer for this. Otherwise we must first add the CSV export functionality, then it will be possible without further requirements.

**Comment by ashimokawa** [comment]
with root access it might be easier to just adb pull the database file.

**Comment by 1337sup3rh4x0r** [comment]
Can't I also just copy it with root Explorer from that location? How shall I send it to you?

**Comment by cpfeiffer** [comment]
Yes, you can. It looks like one can't add attachments here (except for images). So you can either upload it to some public place (dropbox or something like that or you can mail it me, whatever you prefer).

**Comment by cpfeiffer** [comment]
@chrisvella I totally agree with you.

The charting library seems to be a little inflexible regarding the placement of some labels, so it'll take some take to make this really pretty, unless someone comes along and helps :-)

And yes, some hints how to interact with the graphs would really help. Especially regarding zooming x-axis and y-axis (separately).

**Comment by xphnx** [comment]
Here are my db's (FW 1.0.9.3)
[Activitydatabase](https://my.owndrive.com/public.php?service=files&t=29d8e778ba7c7e111035ecf6eb7e36dd)
[ActivityDatabase-journal](https://my.owndrive.com/public.php?service=files&t=9ca1178048034329d7a867bd131dff98)
Maybe would be nice some achievements as in this app https://github.com/j4velin/Pedometer

**Comment by 1337sup3rh4x0r** [comment]
OK, here's my activity dbs:
https://mega.nz/#!2NxyhQ5b!XwelriT8J31LKDWhB7Gzk2LEw1Rn5HCQur_gzmqix_c
https://mega.nz/#!zEgAiJyA!cyNyOYxFdPIfKI8rn0m03IONuETixjBz3BDqjVUqgok

Funny, not wanting to share this data with companies but posting it here :) 

I dont have my miband with me right now. I'll post my Firmware Version later.

**Comment by xphnx** [comment]
The fiels I put previously had no sleep data. Here are the news:
- [ActivityDatabase[(https://my.owndrive.com/public.php?service=files&t=f97fb2a6fc54e7db1f20f0121c1c27ad)
- [ActivityDatabase-journal](https://my.owndrive.com/public.php?service=files&t=4fcda573e8a51ded65d53736da2312ec)
On the other hand perhaps you could enable a mail to send this data in a more convenient way.

**Comment by chrisvella** [comment]
Here are my two files:
http://s000.tinyupload.com/index.php?file_id=62854982291116113158
http://s000.tinyupload.com/index.php?file_id=08034018937859839708

**Comment by cpfeiffer** [comment]
It'll take some time for me to have a look at them, so for the moment, these are enough. Thanks a lot!

**Comment by cpfeiffer** [comment]
@xphnx And yes, we'll get steps tracking soon as well. We're working on it.

**Comment by cpfeiffer** [comment]
@danielegobbetti  I have a problem syncing activity data since a week or so. Well, sync does work, but it never finishes.

This is the last output I get:

```
07-04 11:42:37.634    2088-2105/nodomain.freeyourgadget.gadgetbridge I/nodomain.freeyourgadget.gadgetbridge.miband.MiBandSupport﹕ total data to read: 426 len: 142 minute(s)
07-04 11:42:37.635    2088-2105/nodomain.freeyourgadget.gadgetbridge I/nodomain.freeyourgadget.gadgetbridge.miband.MiBandSupport﹕ data to read until next header: 0 len: 0 minute(s)
07-04 11:42:37.637    2088-2105/nodomain.freeyourgadget.gadgetbridge I/nodomain.freeyourgadget.gadgetbridge.miband.MiBandSupport﹕ TIMESTAMP: 04.07.2015 11:42:24 magic byte: 0
07-04 11:42:37.639    2088-2105/nodomain.freeyourgadget.gadgetbridge D/nodomain.freeyourgadget.gadgetbridge.btle.BtLEQueue﹕ about to add: 11:42:37: Transaction task: send acknowledge with 1 actions
07-04 11:42:37.701    2088-2105/nodomain.freeyourgadget.gadgetbridge D/nodomain.freeyourgadget.gadgetbridge.btle.BtLEQueue﹕ characteristic write: 0000ff05-0000-1000-8000-00805f9b34fb (success)
07-04 11:42:37.702    2088-2105/nodomain.freeyourgadget.gadgetbridge I/nodomain.freeyourgadget.gadgetbridge.miband.MiBandSupport﹕ handleControlPoint got status:0
07-04 11:42:37.704    2088-2105/nodomain.freeyourgadget.gadgetbridge I/nodomain.freeyourgadget.gadgetbridge.miband.MiBandSupport﹕ handleControlPoint GOT DATA:0x       a
07-04 11:42:37.705    2088-2105/nodomain.freeyourgadget.gadgetbridge I/nodomain.freeyourgadget.gadgetbridge.miband.MiBandSupport﹕ handleControlPoint GOT DATA:0x       f
07-04 11:42:37.706    2088-2105/nodomain.freeyourgadget.gadgetbridge I/nodomain.freeyourgadget.gadgetbridge.miband.MiBandSupport﹕ handleControlPoint GOT DATA:0x       6
07-04 11:42:37.707    2088-2105/nodomain.freeyourgadget.gadgetbridge I/nodomain.freeyourgadget.gadgetbridge.miband.MiBandSupport﹕ handleControlPoint GOT DATA:0x       4
07-04 11:42:37.708    2088-2105/nodomain.freeyourgadget.gadgetbridge I/nodomain.freeyourgadget.gadgetbridge.miband.MiBandSupport﹕ handleControlPoint GOT DATA:0x       b
07-04 11:42:37.709    2088-2105/nodomain.freeyourgadget.gadgetbridge I/nodomain.freeyourgadget.gadgetbridge.miband.MiBandSupport﹕ handleControlPoint GOT DATA:0x      28
07-04 11:42:37.710    2088-2105/nodomain.freeyourgadget.gadgetbridge I/nodomain.freeyourgadget.gadgetbridge.miband.MiBandSupport﹕ handleControlPoint GOT DATA:0x      18
07-04 11:42:37.711    2088-2105/nodomain.freeyourgadget.gadgetbridge I/nodomain.freeyourgadget.gadgetbridge.miband.MiBandSupport﹕ handleControlPoint GOT DATA:0x      aa
07-04 11:42:37.712    2088-2105/nodomain.freeyourgadget.gadgetbridge I/nodomain.freeyourgadget.gadgetbridge.miband.MiBandSupport﹕ handleControlPoint GOT DATA:0x       1
07-04 11:42:37.713    2088-2152/nodomain.freeyourgadget.gadgetbridge D/nodomain.freeyourgadget.gadgetbridge.btle.BtLEQueue﹕ About to run action: 11:42:37: WriteAction on characteristic: 0000ff05-0000-1000-8000-00805f9b34fb
07-04 11:42:37.849    2088-2105/nodomain.freeyourgadget.gadgetbridge D/nodomain.freeyourgadget.gadgetbridge.btle.BtLEQueue﹕ characteristic write: 0000ff05-0000-1000-8000-00805f9b34fb (success)
07-04 11:42:37.851    2088-2105/nodomain.freeyourgadget.gadgetbridge I/nodomain.freeyourgadget.gadgetbridge.miband.MiBandSupport﹕ handleControlPoint got status:0
07-04 11:42:37.853    2088-2105/nodomain.freeyourgadget.gadgetbridge I/nodomain.freeyourgadget.gadgetbridge.miband.MiBandSupport﹕ handleControlPoint GOT DATA:0x       a
07-04 11:42:37.854    2088-2105/nodomain.freeyourgadget.gadgetbridge I/nodomain.freeyourgadget.gadgetbridge.miband.MiBandSupport﹕ handleControlPoint GOT DATA:0x       f
07-04 11:42:37.855    2088-2105/nodomain.freeyourgadget.gadgetbridge I/nodomain.freeyourgadget.gadgetbridge.miband.MiBandSupport﹕ handleControlPoint GOT DATA:0x       6
07-04 11:42:37.857    2088-2105/nodomain.freeyourgadget.gadgetbridge I/nodomain.freeyourgadget.gadgetbridge.miband.MiBandSupport﹕ handleControlPoint GOT DATA:0x       4
07-04 11:42:37.859    2088-2105/nodomain.freeyourgadget.gadgetbridge I/nodomain.freeyourgadget.gadgetbridge.miband.MiBandSupport﹕ handleControlPoint GOT DATA:0x       b
07-04 11:42:37.863    2088-2105/nodomain.freeyourgadget.gadgetbridge I/nodomain.freeyourgadget.gadgetbridge.miband.MiBandSupport﹕ handleControlPoint GOT DATA:0x      2a
07-04 11:42:37.866    2088-2105/nodomain.freeyourgadget.gadgetbridge I/nodomain.freeyourgadget.gadgetbridge.miband.MiBandSupport﹕ handleControlPoint GOT DATA:0x      18
07-04 11:42:37.867    2088-2105/nodomain.freeyourgadget.gadgetbridge I/nodomain.freeyourgadget.gadgetbridge.miband.MiBandSupport﹕ handleControlPoint GOT DATA:0x       0
07-04 11:42:37.869    2088-2105/nodomain.freeyourgadget.gadgetbridge I/nodomain.freeyourgadget.gadgetbridge.miband.MiBandSupport﹕ handleControlPoint GOT DATA:0x       0
07-04 11:42:50.369    2088-2088/nodomain.freeyourgadget.gadgetbridge D/nodomain.freeyourgadget.gadgetbridge.BluetoothCommunicationService﹕ Service startcommand: nodomain.freeyourgadget.gadgetbride.bluetoothcommunicationservice.action.start
07-04 11:42:50.370    2088-2088/nodomain.freeyourgadget.gadgetbridge D/nodomain.freeyourgadget.gadgetbridge.BluetoothCommunicationService﹕ Service startcommand: nodomain.freeyourgadget.gadgetbride.bluetoothcommunicationservice.action.request_versioninfo
07-04 11:42:50.402    2088-2088/nodomain.freeyourgadget.gadgetbridge I/nodomain.freeyourgadget.gadgetbridge.GB﹕ Setting broadcast receivers to: true
```

**Comment by cpfeiffer** [comment]
@danielegobbetti Do you happen to know whether we can voluntarily send ACKs during the data transfer? I.e. not just at the end when everything's finished.

Again I have the problem that syncing data of several days starts out good, but then stalls and never finishes.

I'm wondering if the Mi Band has some kind of buffer (of data to be ACKed?) that needs to be flushed before it can continue.

**Comment by danielegobbetti** [comment]
There are two problems that need yo be sooved:

First we don't know when the notification end. Not explicitly I mean. In fact when the notification flow end all the acks that we pipelined are sent, but we would need to intercept this and add another ack.

Second: I don't know why this happens. Usually the number of bytes the miband declares at the beginning of each chunk is the same as the number of bytes we should expect before sending an ack, but in this case it just hangs...

If it's happening again, could you please try performing a READ on the characteristic (instead of the forced ack)? You should be getting 11 bytes, and it would be interesting if it matches the last message you received.


--

**Comment by cpfeiffer** [comment]
Thanks, I'll try that tomorrow or the next day (too late for now). It's reproduceable though, so no problem.

Here's a log + the corresponding hci dump. The dump interestingly contains some data at the end that I have no idea about. 
```
07-14 23:14:59.999  27252-27264/MiBandSupport﹕ total data to read: 12840 len: 4280 minute(s)
07-14 23:15:00.011  27252-27264/MiBandSupport﹕ data to read until next header: 12834 len: 4278 minute(s)
07-14 23:15:00.031  27252-27264/MiBandSupport﹕ TIMESTAMP: 11.07.2015 23:55:53 magic byte: 12834
07-14 23:15:00.034  27252-27264/MiBandSupport﹕ activity data: length: 11, remaining bytes: 12834
07-14 23:15:00.040  27252-27315/ D/BluetoothGatt﹕ onNotify() - Device=88:0F:10:50:C2:45 UUID=0000ff07-0000-1000-8000-00805f9b34fb
[...]
07-14 23:16:00.349  27252-27263/ D/.btle.BtLEQueue﹕ characteristic changed: 0000ff07-0000-1000-8000-00805f9b34fb
07-14 23:16:00.350  27252-27263/MiBandSupport﹕ activity data: length: 20, remaining bytes: 8074
07-14 23:16:00.352  27252-27315/ D/BluetoothGatt﹕ onNotify() - Device=88:0F:10:50:C2:45 UUID=0000ff07-0000-1000-8000-00805f9b34fb
07-14 23:16:00.353  27252-27315/ D/.btle.BtLEQueue﹕ characteristic changed: 0000ff07-0000-1000-8000-00805f9b34fb
07-14 23:16:00.354  27252-27315/MiBandSupport﹕ activity data: length: 20, remaining bytes: 8054
07-14 23:16:00.358  27252-27292/ D/BluetoothGatt﹕ onNotify() - Device=88:0F:10:50:C2:45 UUID=0000ff07-0000-1000-8000-00805f9b34fb
07-14 23:16:00.361  27252-27292/ D/.btle.BtLEQueue﹕ characteristic changed: 0000ff07-0000-1000-8000-00805f9b34fb
07-14 23:16:00.362  27252-27292/MiBandSupport﹕ activity data: length: 20, remaining bytes: 8034
```


```
2015-07-14 23:15:03.202797 > ACL data: handle 15 flags 0x02 dlen 27
    ATT: Handle notify (0x1b)
      handle 0x0020
      value 0x05 0x00 0x00 0x05 0x00 0x00 0x05 0x01 0x00 0x05 0x01 0x00 0x05 0x00 0x00 0x05 0x00 0x00 0x05 0x01 
2015-07-14 23:15:03.263355 > ACL data: handle 15 flags 0x02 dlen 27
    ATT: Handle notify (0x1b)
      handle 0x0020
      value 0x00 0x05 0x00 0x00 0x05 0x01 0x00 0x05 0x04 0x00 0x05 0x03 0x00 0x05 0x03 0x00 0x05 0x03 0x00 0x05 
2015-07-14 23:15:03.322784 > ACL data: handle 15 flags 0x02 dlen 27
    ATT: Handle notify (0x1b)
      handle 0x0020
      value 0x03 0x00 0x05 0x01 0x00 0x05 0x00 0x00 0x05 0x02 0x00 0x05 0x03 0x00 0x05 0x02 0x00 0x05 0x02 0x00 
2015-07-14 23:15:03.382805 > ACL data: handle 15 flags 0x02 dlen 27
    ATT: Handle notify (0x1b)
      handle 0x0020
      value 0x05 0x03 0x00 0x05 0x03 0x00 0x05 0x03 0x00 0x05 0x03 0x00 0x05 0x03 0x00 0x05 0x02 0x00 0x05 0x02 
2015-07-14 23:15:03.442829 > ACL data: handle 15 flags 0x02 dlen 27
    ATT: Handle notify (0x1b)
      handle 0x0020
      value 0x00 0x05 0x01 0x00 0x05 0x02 0x00 0x05 0x01 0x00 0x05 0x00 0x00 0x05 0x02 0x00 0x05 0x02 0x00 0x05 
2015-07-14 23:15:03.443504 > ACL data: handle 15 flags 0x02 dlen 21
    ATT: Handle notify (0x1b)
      handle 0x0020
      value 0x04 0x00 0x05 0x03 0x00 0x05 0x03 0x00 0x05 0x03 0x00 0x05 0x03 0x00 
2015-07-14 23:15:03.445623 > ACL data: handle 15 flags 0x02 dlen 18
    ATT: Handle notify (0x1b)
      handle 0x0020
      value 0x01 0x0f 0x06 0x0e 0x17 0x0d 0x35 0xb8 0x10 0x02 0x00 
2015-07-14 23:15:03.446848 > ACL data: handle 15 flags 0x02 dlen 13
    ATT: Handle notify (0x1b)
      handle 0x0020
      value 0x05 0x03 0x00 0x05 0x03 0x00 
2015-07-14 23:15:03.447717 > ACL data: handle 15 flags 0x02 dlen 18
    ATT: Handle notify (0x1b)
      handle 0x0020
      value 0x01 0x0f 0x06 0x0e 0x17 0x0f 0x35 0xb8 0x10 0x00 0x00 
```

**Comment by cpfeiffer** [comment]
BTW, I tried sending an ACK in fliushActivityDataHolder(), but to no avail.

**Comment by danielegobbetti** [comment]
You aren't anyway sending the acks where you queue them :)

As far ad I can tell from the data in the hci dump these data at the end are normal notification with length 20 that (at first sight) look like sleep data (category 5, intensity 0 or 1, steps 0)... I wonder why the BT stack isn't passing it up to us.


--
